core/src/main/java/org/pentaho/cassandra/driver/datastax/DriverCQLRowHandler.java

```diff
@@ -152,22 +152,59 @@ public void newRowQuery( StepInterface requestingStep, String tableName, String
   }
 
   public static Object readValue( ValueMetaInterface meta, Row row, int i ) {
+    
+    if (row.isNull( i ) ) {
+      return null;
+    }
+    
     switch ( meta.getType() ) {
+      
       case ValueMetaInterface.TYPE_INTEGER:
-        return row.getLong( i );
       case ValueMetaInterface.TYPE_NUMBER:
-        return row.getDouble( i );
       case ValueMetaInterface.TYPE_BIGNUMBER:
-        return row.getDecimal( i );
+      case ValueMetaInterface.TYPE_BOOLEAN:
       case ValueMetaInterface.TYPE_DATE:
-        // Check whether this is a CQL Date or Timestamp
+      case ValueMetaInterface.TYPE_BINARY:
+        /*
+         * https://docs.datastax.com/en/cql/3.3/cql/cql_reference/cql_data_types_c.html
+         * 
+         * Calling getLong() on a Cassandra column of INT datatype throws an NPE
+         * Calling getDouble() on a Cassandra column of FLOAT datatype throws an NPE
+         * ...
+         */
         ColumnDefinitions cdef = row.getColumnDefinitions();
-        if ( cdef.getType( i ).getName() == DataType.Name.DATE ) {
-          LocalDate ld = row.getDate( i );
-          return new java.util.Date(  ld.getMillisSinceEpoch() );
-        } else {
-          return row.getTimestamp( i );
+        switch( cdef.getType( i ).getName() ) {
+          case BIGINT:
+          case COUNTER:
+            return row.getLong( i );
+          case SMALLINT:
+            return new Long( row.getShort( i ) );
+          case TINYINT:
+            return new Long( row.getByte( i ) );
```

Comment:
I cannot see why a TINYINT should return a Long. That makes no sense to me. JDBC TINYINT is an 8-bit unsigned int between 0 and 255 ( 00-FF ) - basically a byte. Suggest this should be using a byte. Clearly, you're using row.getByte, but it's unclear to me why you would return a new Byte ...